g3doc/Tutorials/Toy/Ch-2.md

```diff
@@ -23,139 +23,320 @@ addresses this issue by being designed for extensibility. As such, there are
+                      /// The ConstantOp takes zero inputs.
+                      mlir::OpTrait::ZeroOperands,
+                      /// The ConstantOp returns a single result.
+                      mlir::OpTrait::OneResult> {
+
+ public:
+  /// Inherit the constructors from the base Op class.
+  using Op::Op;
+
+  /// Provide the unique name for this operation. MLIR will use this to register
+  /// the operation and uniquely identify it throughout the system.
+  static llvm::StringRef getOperationName() { return "toy.constant"; }
+
+  /// Operations can provide additional verification beyond the traits they
+  /// define. Here we will ensure that the specific invariants of the constant
+  /// operation are upheld, for example the result type must be of TensorType.
+  LogicalResult verify();
+
+  /// Provide an interface to build this operation from a set of input values.
+  /// This interface is used by the builder to allow for easily generating
+  /// instances of this operation:
+  ///   mlir::OpBuilder::create<ConstantOp>(...)
+  /// This method populates the given `state` that MLIR uses to create
+  /// operations. This state is a collection of all of the discrete elements
+  /// that an operation may contain.
+  /// Build a constant with the given return type and `value` attribute.
+  static void build(mlir::Builder *builder, mlir::OperationState *state,
+                    mlir::Type result, mlir::DenseElementsAttr value);
+  /// Build a constant and reuse the type from the given 'value'.
+  static void build(mlir::Builder *builder, mlir::OperationState *state,
+                    mlir::DenseElementsAttr value);
+  /// Build a constant by broadcasting the given 'value'.
+  static void build(mlir::Builder *builder, mlir::OperationState *state,
+                    double value);
+
+  /// Return the value of the constant by fetching it from the attribute.
+  mlir::DenseElementsAttr getValue();
+};
 ```
 
-This approach is used in `Ch2/mlir/MLIRGen.cpp` to implement a naive MLIR
-generation through a simple depth-first search traversal of the Toy AST. Here is
-how we create a `toy.transpose` operation:
+and we register this operation in the `ToyDialect` constructor:
 
 ```c++
-mlir::Operation *createTransposeOp(OpBuilder &builder,
-                                   mlir::Value *input_tensor) {
-  // Fill the `OperationState` with the required fields.
-  mlir::OperationState result(location, "toy.transpose");
-  result.addOperands(input_tensor);
-
-  // We use the MLIR tensor type for 'toy' types.
-  auto type = builder.getTensorType({2, 2}, builder.getF64Type());
-  result.addTypes(type);
-
-  // Create the transpose operation.
-  Operation *newTransposeOp = builder->createOperation(result);
-  return newTransposeOp;
+ToyDialect::ToyDialect(mlir::MLIRContext *ctx)
+    : mlir::Dialect(getDialectNamespace(), ctx) {
+  addOperations<ConstantOp>();
+}
+```
+
+### Op vs Operation: Using MLIR Operations
```

Comment:
did we brainstorm naming recently? I don't remember if we concluded anything. What about renaming Operation into OpaqueOp for instance?